### DISCUSSION TITLE
Idea on NNUE speedup

### DISCUSSION THREAD

**Discussion by cj5716** [official-stockfish/Stockfish#5180] [Ideas]
There is a possibility that we can speedup refreshes by using something called a "Finny table", introduced by Luecx (one of the authors of Koivisto). A relevant commit: https://github.com/AndyGrant/Ethereal/commit/c5ba24dfcfd46c87f1c6087bb151d92beacacb5b (note that this is in Ethereal instead)
The way this works is that we have an array of computed accumulators + bitboards indexed by bucket index. This way, when we require a refresh, we can simply update the difference from the computed accumulator rather than from scratch.
In addition, Andrew has been speaking quite a bit on potential NNUE speedups that Torch has but Stockfish doesn't, and there is a high chance that this is one of them.
So far, I have checked and it seems Stockfish doesn't have this idea. Maybe we can explore this? I am not sure how easy this is to implement.
cc @mstembera I think you might be interested :)

**Comment by Sopel97** [comment]
That's clever. Should be strictly better indeed, with a cost of about 500kB of RAM per thread in the current network architecture.

**Reply by AndyGrant** [reply]
Make sure you credit Luecx when this invariably passes all testing as a massive speed gain.

**Comment by cj5716** [comment]
complete with https://github.com/official-stockfish/Stockfish/commit/49ef4c935a5cb0e4d94096e6354caa06b36b3e3c
